// ngx-easy-table/bundles/ngx-easy-table.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('rxjs'), require('rxjs/operators'), require('@angular/cdk/drag-drop'), require('@angular/common'), require('@angular/core'), require('ngx-pagination')) :
    typeof define === 'function' && define.amd ? define('ngx-easy-table', ['exports', 'rxjs', 'rxjs/operators', '@angular/cdk/drag-drop', '@angular/common', '@angular/core', 'ngx-pagination'], factory) :
    (factory((global['ngx-easy-table'] = {}),global.rxjs,global.rxjs.operators,global.ng.cdk['drag-drop'],global.ng.common,global.ng.core,global.ngxPagination));
}(this, (function (exports,rxjs,operators,dragDrop,common,core,ngxPagination) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var STYLE = {
        TINY: 'tiny',
        BIG: 'big',
        NORMAL: 'normal',
    };
    /** @enum {string} */
    var THEME = {
        LIGHT: 'light',
        DARK: 'dark',
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var Event = {
        onPagination: 'onPagination',
        onOrder: 'onOrder',
        onGlobalSearch: 'onGlobalSearch',
        onSearch: 'onSearch',
        onClick: 'onClick',
        onDoubleClick: 'onDoubleClick',
        onCheckboxSelect: 'onCheckboxSelect',
        onSelectAll: 'onSelectAll',
        onColumnResizeMouseDown: 'onColumnResizeMouseDown',
        onColumnResizeMouseUp: 'onColumnResizeMouseUp',
        onRowDrop: 'onRowDrop',
        onRowCollapsedShow: 'onRowCollapsedShow',
        onRowCollapsedHide: 'onRowCollapsedHide',
        onRowContextMenu: 'onRowContextMenu',
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @enum {string} */
    var API = {
        rowContextMenuClicked: 'rowContextMenuClicked',
        setInputValue: 'setInputValue',
        toolPanelClicked: 'toolPanelClicked',
        onGlobalSearch: 'onGlobalSearch',
        setPaginationCurrentPage: 'setPaginationCurrentPage',
        getPaginationCurrentPage: 'getPaginationCurrentPage',
        getPaginationTotalItems: 'getPaginationTotalItems',
        getPaginationLastPage: 'getPaginationLastPage',
        setPaginationRange: 'setPaginationRange',
        setPaginationPreviousLabel: 'setPaginationPreviousLabel',
        setPaginationNextLabel: 'setPaginationNextLabel',
        setPaginationDisplayLimit: 'setPaginationDisplayLimit',
        setTableClass: 'setTableClass',
        setRowClass: 'setRowClass',
        setCellClass: 'setCellClass',
        setRowStyle: 'setRowStyle',
        setCellStyle: 'setCellStyle',
        sortBy: 'sortBy',
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var DefaultConfig = {
        searchEnabled: false,
        headerEnabled: true,
        orderEnabled: true,
        orderEventOnly: false,
        paginationEnabled: true,
        exportEnabled: false,
        clickEvent: true,
        selectRow: false,
        selectCol: false,
        selectCell: false,
        rows: 10,
        additionalActions: false,
        serverPagination: false,
        isLoading: false,
        detailsTemplate: false,
        groupRows: false,
        paginationRangeEnabled: true,
        collapseAllRows: false,
        checkboxes: false,
        resizeColumn: false,
        fixedColumnWidth: true,
        horizontalScroll: false,
        draggable: false,
        logger: false,
        showDetailsArrow: false,
        showContextMenu: false,
        persistState: false,
        paginationMaxSize: 5,
        threeWaySort: false,
        tableLayout: {
            style: STYLE.NORMAL,
            theme: THEME.LIGHT,
            borderless: false,
            hover: true,
            striped: false,
        },
    };
    var DefaultConfigService = /** @class */ (function () {
        function DefaultConfigService() {
        }
        DefaultConfigService.config = DefaultConfig;
        DefaultConfigService.decorators = [
            { type: core.Injectable }
        ];
        return DefaultConfigService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var GroupRowsService = /** @class */ (function () {
        function GroupRowsService() {
        }
        /**
         * @template T
         * @param {?} data
         * @param {?} groupRowsBy
         * @return {?}
         */
        GroupRowsService.doGroupRows = /**
         * @template T
         * @param {?} data
         * @param {?} groupRowsBy
         * @return {?}
         */
            function (data, groupRowsBy) {
                /** @type {?} */
                var grouped = [];
                rxjs.from(data).pipe(operators.groupBy(( /**
                 * @param {?} row
                 * @return {?}
                 */function (row) { return row[groupRowsBy]; })), operators.flatMap(( /**
                 * @param {?} group
                 * @return {?}
                 */function (group) {
                    return group.pipe(operators.reduce(( /**
                     * @param {?} acc
                     * @param {?} curr
                     * @return {?}
                     */function (acc, curr) { return __spread(acc, [curr]); }), []));
                }))).subscribe(( /**
                 * @param {?} row
                 * @return {?}
                 */function (row) { return grouped.push(row); }));
                return grouped;
            };
        GroupRowsService.decorators = [
            { type: core.Injectable }
        ];
        return GroupRowsService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var StyleService = /** @class */ (function () {
        function StyleService() {
        }
        /**
         * @param {?} val
         * @return {?}
         */
        StyleService.setRowClass = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                /** @type {?} */
                var selector = "#table > tbody > tr:nth-child(" + val.row + ")";
                /** @type {?} */
                var row = document.querySelector(selector);
                if (row) {
                    row.className = val.className;
                }
            };
        /**
         * @param {?} val
         * @return {?}
         */
        StyleService.setCellClass = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                /** @type {?} */
                var selector = "#table > tbody > tr:nth-child(" + val.row + ") > td:nth-child(" + val.cell + ")";
                /** @type {?} */
                var cell = document.querySelector(selector);
                if (cell) {
                    cell.className = val.className;
                }
            };
        /**
         * @param {?} val
         * @return {?}
         */
        StyleService.setRowStyle = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                /** @type {?} */
                var selector = "#table > tbody > tr:nth-child(" + val.row + ")";
                /** @type {?} */
                var row = document.querySelector(selector);
                if (row) {
                    // tslint:disable-next-line:no-string-literal
                    row.style[val.attr] = val.value;
                }
            };
        /**
         * @param {?} val
         * @return {?}
         */
        StyleService.setCellStyle = /**
         * @param {?} val
         * @return {?}
         */
            function (val) {
                /** @type {?} */
                var selector = "#table > tbody > tr:nth-child(" + val.row + ") > td:nth-child(" + val.cell + ")";
                /** @type {?} */
                var cell = document.querySelector(selector);
                if (cell) {
                    // tslint:disable-next-line:no-string-literal
                    cell.style[val.attr] = val.value;
                }
            };
        StyleService.decorators = [
            { type: core.Injectable }
        ];
        return StyleService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BaseComponent = /** @class */ (function () {
        function BaseComponent(cdr) {
            this.cdr = cdr;
            this.tableClass = null;
            this.grouped = [];
            this.menuActive = false;
            this.isSelected = false;
            this.page = 1;
            this.count = null;
            this.sortState = new Map();
            this.sortKey = null;
            this.rowContextMenuPosition = {
                top: null,
                left: null,
                value: null,
            };
            this.sortBy = {
                key: '',
                order: 'asc',
            };
            this.selectedDetailsTemplateRowId = new Set();
            this.loadingHeight = '30px';
            this.onSelectAllBinded = this.onSelectAll.bind(this);
            this.id = 'table';
            this.event = new core.EventEmitter();
        }
        Object.defineProperty(BaseComponent.prototype, "configuration", {
            get: /**
             * @return {?}
             */ function () {
                return this.config;
            },
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this.config = value;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        BaseComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                if (!this.columns) {
                    console.error('[columns] property required!');
                }
                if (this.configuration) {
                    DefaultConfigService.config = this.configuration;
                }
                this.config = DefaultConfigService.config;
                this.limit = this.config.rows;
                if (this.groupRowsBy) {
                    this.grouped = GroupRowsService.doGroupRows(this.data, this.groupRowsBy);
                }
                this.doDecodePersistedState();
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        BaseComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                /** @type {?} */
                var data = changes.data;
                /** @type {?} */
                var pagination = changes.pagination;
                /** @type {?} */
                var groupRowsBy = changes.groupRowsBy;
                this.toggleRowIndex = changes.toggleRowIndex;
                if (data && data.currentValue) {
                    this.doApplyData(data);
                }
                if (pagination && pagination.currentValue) {
                    this.count = pagination.currentValue.count;
                }
                if (groupRowsBy && groupRowsBy.currentValue) {
                    this.grouped = GroupRowsService.doGroupRows(this.data, this.groupRowsBy);
                }
                if (this.toggleRowIndex && this.toggleRowIndex.currentValue) {
                    /** @type {?} */
                    var row = this.toggleRowIndex.currentValue;
                    this.collapseRow(row.index);
                }
            };
        /**
         * @param {?} column
         * @return {?}
         */
        BaseComponent.prototype.isOrderEnabled = /**
         * @param {?} column
         * @return {?}
         */
            function (column) {
                /** @type {?} */
                var columnOrderEnabled = column.orderEnabled === undefined ? true : !!column.orderEnabled;
                return DefaultConfigService.config.orderEnabled && columnOrderEnabled;
            };
        /**
         * @param {?} column
         * @return {?}
         */
        BaseComponent.prototype.orderBy = /**
         * @param {?} column
         * @return {?}
         */
            function (column) {
                if (typeof column.orderEnabled !== 'undefined' && !column.orderEnabled) {
                    return;
                }
                this.sortKey = column.key;
                if (!DefaultConfigService.config.orderEnabled || this.sortKey === '') {
                    return;
                }
                this.setColumnOrder(this.sortKey);
                if (!DefaultConfigService.config.orderEventOnly && !column.orderEventOnly) {
                    this.sortBy.key = this.sortKey;
                    this.sortBy.order = this.sortState.get(this.sortKey);
                }
                else {
                    this.sortBy.key = '';
                    this.sortBy.order = '';
                }
                if (!DefaultConfigService.config.serverPagination) {
                    this.data = __spread(this.data);
                }
                this.sortBy = __assign({}, this.sortBy);
                /** @type {?} */
                var value = {
                    key: this.sortKey,
                    order: this.sortState.get(this.sortKey),
                };
                this.emitEvent(Event.onOrder, value);
            };
        /**
         * @param {?} $event
         * @param {?} row
         * @param {?} key
         * @param {?} colIndex
         * @param {?} rowIndex
         * @return {?}
         */
        BaseComponent.prototype.onClick = /**
         * @param {?} $event
         * @param {?} row
         * @param {?} key
         * @param {?} colIndex
         * @param {?} rowIndex
         * @return {?}
         */
            function ($event, row, key, colIndex, rowIndex) {
                if (DefaultConfigService.config.selectRow) {
                    this.selectedRow = rowIndex;
                }
                if (DefaultConfigService.config.selectCol && colIndex) {
                    this.selectedCol = colIndex;
                }
                if (DefaultConfigService.config.selectCell && colIndex) {
                    this.selectedRow = rowIndex;
                    this.selectedCol = colIndex;
                }
                if (DefaultConfigService.config.clickEvent) {
                    /** @type {?} */
                    var value = {
                        event: $event,
                        row: row,
                        key: key,
                        rowId: rowIndex,
                        colId: colIndex,
                    };
                    this.emitEvent(Event.onClick, value);
                }
            };
        /**
         * @param {?} $event
         * @param {?} row
         * @param {?} key
         * @param {?} colIndex
         * @param {?} rowIndex
         * @return {?}
         */
        BaseComponent.prototype.onDoubleClick = /**
         * @param {?} $event
         * @param {?} row
         * @param {?} key
         * @param {?} colIndex
         * @param {?} rowIndex
         * @return {?}
         */
            function ($event, row, key, colIndex, rowIndex) {
                /** @type {?} */
                var value = {
                    event: $event,
                    row: row,
                    key: key,
                    rowId: rowIndex,
                    colId: colIndex,
                };
                this.emitEvent(Event.onDoubleClick, value);
            };
        /**
         * @param {?} $event
         * @param {?} row
         * @param {?} rowIndex
         * @return {?}
         */
        BaseComponent.prototype.onCheckboxSelect = /**
         * @param {?} $event
         * @param {?} row
         * @param {?} rowIndex
         * @return {?}
         */
            function ($event, row, rowIndex) {
                /** @type {?} */
                var value = {
                    event: $event,
                    row: row,
                    rowId: rowIndex,
                };
                this.emitEvent(Event.onCheckboxSelect, value);
            };
        /**
         * @return {?}
         */
        BaseComponent.prototype.onSelectAll = /**
         * @return {?}
         */
            function () {
                this.isSelected = !this.isSelected;
                this.emitEvent(Event.onSelectAll, this.isSelected);
            };
        /**
         * @param {?} $event
         * @return {?}
         */
        BaseComponent.prototype.onSearch = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                if (!DefaultConfigService.config.serverPagination) {
                    this.term = $event;
                }
                this.emitEvent(Event.onSearch, $event);
            };
        /**
         * @param {?} value
         * @return {?}
         */
        BaseComponent.prototype.onGlobalSearch = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                if (!DefaultConfigService.config.serverPagination) {
                    this.globalSearchTerm = value;
                }
                this.emitEvent(Event.onGlobalSearch, value);
            };
        /**
         * @param {?} pagination
         * @return {?}
         */
        BaseComponent.prototype.onPagination = /**
         * @param {?} pagination
         * @return {?}
         */
            function (pagination) {
                this.page = pagination.page;
                this.limit = pagination.limit;
                this.emitEvent(Event.onPagination, pagination);
            };
        /**
         * @private
         * @param {?} event
         * @param {?} value
         * @return {?}
         */
        BaseComponent.prototype.emitEvent = /**
         * @private
         * @param {?} event
         * @param {?} value
         * @return {?}
         */
            function (event, value) {
                this.event.emit({ event: event, value: value });
                if (this.config.persistState) {
                    localStorage.setItem(event, JSON.stringify(value));
                }
                if (this.config.logger) {
                    // tslint:disable-next-line:no-console
                    console.log({ event: event, value: value });
                }
            };
        /**
         * @param {?} rowIndex
         * @return {?}
         */
        BaseComponent.prototype.collapseRow = /**
         * @param {?} rowIndex
         * @return {?}
         */
            function (rowIndex) {
                if (this.selectedDetailsTemplateRowId.has(rowIndex)) {
                    this.selectedDetailsTemplateRowId.delete(rowIndex);
                    this.emitEvent(Event.onRowCollapsedHide, rowIndex);
                }
                else {
                    this.selectedDetailsTemplateRowId.add(rowIndex);
                    this.emitEvent(Event.onRowCollapsedShow, rowIndex);
                }
            };
        /**
         * @private
         * @return {?}
         */
        BaseComponent.prototype.doDecodePersistedState = /**
         * @private
         * @return {?}
         */
            function () {
                if (!this.config.persistState) {
                    return;
                }
                /** @type {?} */
                var pagination = localStorage.getItem(Event.onPagination);
                /** @type {?} */
                var sort = localStorage.getItem(Event.onOrder);
                /** @type {?} */
                var search = localStorage.getItem(Event.onSearch);
                if (pagination) {
                    this.onPagination(JSON.parse(pagination));
                }
                if (sort) {
                    var _a = JSON.parse(sort), key = _a.key, order = _a.order;
                    this.sortBy.key = key;
                    this.sortBy.order = order;
                    this.data = __spread(this.data);
                }
                if (search) {
                    this.term = JSON.parse(search);
                }
            };
        /**
         * @param {?} rowIndex
         * @return {?}
         */
        BaseComponent.prototype.isRowCollapsed = /**
         * @param {?} rowIndex
         * @return {?}
         */
            function (rowIndex) {
                if (this.config.collapseAllRows) {
                    return true;
                }
                return this.selectedDetailsTemplateRowId.has(rowIndex);
            };
        /**
         * @param {?} event
         * @param {?} th
         * @return {?}
         */
        BaseComponent.prototype.onMouseDown = /**
         * @param {?} event
         * @param {?} th
         * @return {?}
         */
            function (event, th) {
                if (!this.config.resizeColumn) {
                    return;
                }
                this.th = th;
                this.startOffset = th.offsetWidth - event.pageX;
                this.emitEvent(Event.onColumnResizeMouseDown, event);
            };
        /**
         * @param {?} event
         * @return {?}
         */
        BaseComponent.prototype.onMouseMove = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (!this.config.resizeColumn) {
                    return;
                }
                if (this.th && this.th.style) {
                    this.th.style.width = this.startOffset + event.pageX + 'px';
                    this.th.style.cursor = 'col-resize';
                    this.th.style['user-select'] = 'none';
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        BaseComponent.prototype.onMouseUp = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (!this.config.resizeColumn) {
                    return;
                }
                this.emitEvent(Event.onColumnResizeMouseUp, event);
                this.th.style.cursor = 'default';
                this.th = undefined;
            };
        Object.defineProperty(BaseComponent.prototype, "isLoading", {
            get: /**
             * @return {?}
             */ function () {
                /** @type {?} */
                var table = ( /** @type {?} */(document.getElementById('table')));
                if (table && table.rows && table.rows.length > 3) {
                    this.getLoadingHeight(table.rows);
                }
                return this.config.isLoading;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} rows
         * @return {?}
         */
        BaseComponent.prototype.getLoadingHeight = /**
         * @param {?} rows
         * @return {?}
         */
            function (rows) {
                /** @type {?} */
                var searchEnabled = this.config.searchEnabled ? 1 : 0;
                /** @type {?} */
                var headerEnabled = this.config.headerEnabled ? 1 : 0;
                /** @type {?} */
                var borderTrHeight = 1;
                /** @type {?} */
                var borderDivHeight = 2;
                this.loadingHeight = (rows.length - searchEnabled - headerEnabled) * (rows[3].offsetHeight - borderTrHeight) - borderDivHeight + 'px';
            };
        /**
         * @param {?} column
         * @return {?}
         */
        BaseComponent.prototype.getColumnWidth = /**
         * @param {?} column
         * @return {?}
         */
            function (column) {
                if (column.width) {
                    return column.width;
                }
                return this.config.fixedColumnWidth ? 100 / this.columns.length + '%' : null;
            };
        /**
         * @param {?} column
         * @return {?}
         */
        BaseComponent.prototype.getColumnDefinition = /**
         * @param {?} column
         * @return {?}
         */
            function (column) {
                return column.searchEnabled || typeof column.searchEnabled === 'undefined';
            };
        Object.defineProperty(BaseComponent.prototype, "arrowDefinition", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.showDetailsArrow || typeof this.config.showDetailsArrow === 'undefined';
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} $event
         * @param {?} row
         * @param {?} key
         * @param {?} colIndex
         * @param {?} rowIndex
         * @return {?}
         */
        BaseComponent.prototype.onRowContextMenu = /**
         * @param {?} $event
         * @param {?} row
         * @param {?} key
         * @param {?} colIndex
         * @param {?} rowIndex
         * @return {?}
         */
            function ($event, row, key, colIndex, rowIndex) {
                if (!this.config.showContextMenu) {
                    return;
                }
                $event.preventDefault();
                /** @type {?} */
                var value = {
                    event: $event,
                    row: row,
                    key: key,
                    rowId: rowIndex,
                    colId: colIndex,
                };
                this.rowContextMenuPosition = {
                    top: $event.y - 10 + "px",
                    left: $event.x - 10 + "px",
                    value: value,
                };
                this.emitEvent(Event.onRowContextMenu, value);
            };
        /**
         * @private
         * @param {?} data
         * @return {?}
         */
        BaseComponent.prototype.doApplyData = /**
         * @private
         * @param {?} data
         * @return {?}
         */
            function (data) {
                /** @type {?} */
                var column = this.columns.find(( /**
                 * @param {?} c
                 * @return {?}
                 */function (c) { return !!c.orderBy; }));
                if (column) {
                    this.sortState.set(this.sortKey, (column.orderBy === 'asc') ? 'desc' : 'asc');
                    this.orderBy(column);
                }
                else {
                    this.data = __spread(data.currentValue);
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        BaseComponent.prototype.onDrop = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.emitEvent(Event.onRowDrop, event);
                dragDrop.moveItemInArray(this.data, event.previousIndex, event.currentIndex);
            };
        // DO NOT REMOVE. It is called from parent component. See src/app/demo/api-doc/api-doc.component.ts
        // DO NOT REMOVE. It is called from parent component. See src/app/demo/api-doc/api-doc.component.ts
        /**
         * @param {?} event
         * @return {?}
         */
        BaseComponent.prototype.apiEvent =
            // DO NOT REMOVE. It is called from parent component. See src/app/demo/api-doc/api-doc.component.ts
            /**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                return this.bindApi(event);
            };
        // tslint:disable:no-big-function cognitive-complexity
        // tslint:disable:no-big-function cognitive-complexity
        /**
         * @private
         * @param {?} event
         * @return {?}
         */
        BaseComponent.prototype.bindApi =
            // tslint:disable:no-big-function cognitive-complexity
            /**
             * @private
             * @param {?} event
             * @return {?}
             */
            function (event) {
                switch (event.type) {
                    case API.rowContextMenuClicked:
                        this.rowContextMenuPosition = {
                            top: null,
                            left: null,
                            value: null,
                        };
                        break;
                    case API.toolPanelClicked:
                        // TODO
                        break;
                    case API.setInputValue:
                        if (this.config.searchEnabled) {
                            event.value.forEach(( /**
                             * @param {?} input
                             * @return {?}
                             */function (input) {
                                /** @type {?} */
                                var element = (( /** @type {?} */(document.getElementById("search_" + input.key))));
                                if (!element) {
                                    console.error("Column '" + input.key + "' not available in the DOM. Have you misspelled a name?");
                                }
                                else {
                                    element.value = input.value;
                                }
                            }));
                        }
                        this.onSearch(event.value);
                        this.cdr.detectChanges();
                        break;
                    case API.onGlobalSearch:
                        this.onGlobalSearch(event.value);
                        this.cdr.detectChanges();
                        break;
                    case API.setRowClass:
                        if (Array.isArray(event.value)) {
                            event.value.forEach(( /**
                             * @param {?} val
                             * @return {?}
                             */function (val) { return StyleService.setRowClass(val); }));
                            break;
                        }
                        StyleService.setRowClass(event.value);
                        this.cdr.detectChanges();
                        break;
                    case API.setCellClass:
                        if (Array.isArray(event.value)) {
                            event.value.forEach(( /**
                             * @param {?} val
                             * @return {?}
                             */function (val) { return StyleService.setCellClass(val); }));
                            break;
                        }
                        StyleService.setCellClass(event.value);
                        break;
                    case API.setRowStyle:
                        if (Array.isArray(event.value)) {
                            event.value.forEach(( /**
                             * @param {?} val
                             * @return {?}
                             */function (val) { return StyleService.setRowStyle(val); }));
                            break;
                        }
                        StyleService.setRowStyle(event.value);
                        break;
                    case API.setCellStyle:
                        if (Array.isArray(event.value)) {
                            event.value.forEach(( /**
                             * @param {?} val
                             * @return {?}
                             */function (val) { return StyleService.setCellStyle(val); }));
                            break;
                        }
                        StyleService.setCellStyle(event.value);
                        break;
                    case API.setTableClass:
                        this.tableClass = event.value;
                        this.cdr.detectChanges();
                        break;
                    case API.getPaginationTotalItems:
                        return this.paginationComponent.paginationDirective.getTotalItems();
                    case API.setPaginationCurrentPage:
                        this.paginationComponent.paginationDirective.setCurrent(event.value);
                        break;
                    case API.setPaginationRange:
                        this.paginationComponent.ranges = event.value;
                        break;
                    case API.setPaginationPreviousLabel:
                        this.paginationComponent.previousLabel = event.value;
                        break;
                    case API.setPaginationNextLabel:
                        this.paginationComponent.nextLabel = event.value;
                        break;
                    case API.setPaginationDisplayLimit:
                        this.paginationComponent.changeLimit(event.value, true);
                        break;
                    case API.sortBy:
                        /** @type {?} */
                        var column = { title: '', key: event.value.column, orderBy: event.value.order };
                        this.orderBy(column);
                        this.cdr.detectChanges();
                        break;
                    default:
                        break;
                }
            };
        /**
         * @private
         * @param {?} key
         * @return {?}
         */
        BaseComponent.prototype.setColumnOrder = /**
         * @private
         * @param {?} key
         * @return {?}
         */
            function (key) {
                switch (this.sortState.get(key)) {
                    case '':
                    case undefined:
                        this.sortState.set(key, 'desc');
                        break;
                    case 'asc':
                        this.config.threeWaySort ?
                            this.sortState.set(key, '') :
                            this.sortState.set(key, 'desc');
                        break;
                    case 'desc':
                        this.sortState.set(key, 'asc');
                        break;
                }
                if (this.sortState.size > 1) {
                    /** @type {?} */
                    var temp = this.sortState.get(key);
                    this.sortState.clear();
                    this.sortState.set(key, temp);
                }
            };
        BaseComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ngx-table',
                        providers: [DefaultConfigService, GroupRowsService, StyleService],
                        template: "<div class=\"ngx-container\">\n  <table [id]=\"id\"\n         [ngClass]=\"(tableClass === null || tableClass === '') ? 'ngx-table' : tableClass\"\n         [class.ngx-table__table--tiny]=\"config.tableLayout.style === 'tiny'\"\n         [class.ngx-table__table--normal]=\"config.tableLayout.style === 'normal'\"\n         [class.ngx-table__table--big]=\"config.tableLayout.style === 'big'\"\n         [class.ngx-table__table--borderless]=\"config.tableLayout.borderless\"\n         [class.ngx-table__table--dark]=\"config.tableLayout.theme === 'dark'\"\n         [class.ngx-table__table--hoverable]=\"config.tableLayout.hover\"\n         [class.ngx-table__table--striped]=\"config.tableLayout.striped\"\n         [class.ngx-table__horizontal-scroll]=\"config.horizontalScroll && !isLoading\">\n    <thead>\n    <tr class=\"ngx-table__header\" *ngIf=\"config.headerEnabled\">\n      <th *ngIf=\"config.checkboxes\" width=\"3%\">\n        <ng-container\n          *ngIf=\"selectAllTemplate\"\n          [ngTemplateOutlet]=\"selectAllTemplate\"\n          [ngTemplateOutletContext]=\"{ $implicit: onSelectAllBinded }\">\n        </ng-container>\n        <label class=\"ngx-form-checkbox\" *ngIf=\"!selectAllTemplate\">\n          <input type=\"checkbox\" (change)=\"onSelectAll()\">\n          <i class=\"ngx-form-icon\" id=\"selectAllCheckbox\"></i>\n        </label>\n      </th>\n      <ng-container *ngFor=\"let column of columns;\">\n        <th class=\"ngx-table__header-cell\"\n            #th\n            [attr.width]=\"getColumnWidth(column)\"\n            (mousedown)=\"onMouseDown($event, th)\"\n            (mouseup)=\"onMouseUp($event)\"\n            (mousemove)=\"onMouseMove($event)\">\n          <div (click)=\"orderBy(column)\" [class.pointer]=\"isOrderEnabled(column)\">\n            <div class=\"ngx-table__header-title\">{{ column.title }}<span>&nbsp;</span>\n              <div [style.display]=\"config.orderEnabled ? 'inline' : 'none' \">\n                <span *ngIf=\"sortKey === column.key && this.sortState.get(sortKey) === ''\"> -</span>\n                <i *ngIf=\"sortKey === column.key && this.sortState.get(sortKey) === 'asc'\"\n                   class=\"ngx-icon ngx-icon-arrow-up\">\n                </i>\n                <i *ngIf=\"sortKey === column.key && this.sortState.get(sortKey) === 'desc'\"\n                   class=\"ngx-icon ngx-icon-arrow-down\">\n                </i>\n              </div>\n            </div>\n          </div>\n          <div class=\"ngx-table__column-resizer\" *ngIf=\"config.resizeColumn\"></div>\n        </th>\n      </ng-container>\n      <th *ngIf=\"config.additionalActions || config.detailsTemplate || config.collapseAllRows || config.groupRows\"\n          class=\"ngx-table__header-cell-additional-actions\">\n        <div class=\"ngx-dropdown ngx-active\"\n             *ngIf=\"config.additionalActions\"\n             [class.ngx-active]=\"menuActive\">\n          <a class=\"ngx-btn ngx-btn-link\" (click)=\"menuActive = !menuActive\">\n            <span class=\"ngx-icon ngx-icon-menu\"></span>\n          </a>\n          <ul class=\"ngx-menu ngx-table__table-menu\">\n            <li class=\"ngx-menu-item\">\n              <app-csv-export [data]=\"data\"\n                              *ngIf=\"config.exportEnabled\">\n              </app-csv-export>\n            </li>\n          </ul>\n        </div>\n      </th>\n    </tr>\n    <tr *ngIf=\"config.searchEnabled && !filtersTemplate\"\n        class=\"ngx-table__sort-header\">\n      <th *ngIf=\"config.checkboxes\"></th>\n      <ng-container *ngFor=\"let column of columns\">\n        <th>\n          <table-header\n            *ngIf=\"getColumnDefinition(column)\"\n            (update)=\"onSearch($event)\"\n            [column]=\"column\">\n          </table-header>\n        </th>\n      </ng-container>\n      <th *ngIf=\"config.additionalActions || config.detailsTemplate\"></th>\n    </tr>\n    <ng-container *ngIf=\"filtersTemplate\">\n      <tr>\n        <ng-container [ngTemplateOutlet]=\"filtersTemplate\">\n        </ng-container>\n      </tr>\n    </ng-container>\n    </thead>\n    <tbody *ngIf=\"data && !isLoading  && !config.draggable\">\n    <ng-container *ngIf=\"rowTemplate\">\n      <ul class=\"ngx-table__table-row-context-menu\"\n          [ngStyle]=\"{'position': 'absolute', 'top': rowContextMenuPosition.top, 'left': rowContextMenuPosition.left }\"\n          *ngIf=\"rowContextMenuPosition.top\">\n        <ng-container\n          [ngTemplateOutlet]=\"rowContextMenu\"\n          [ngTemplateOutletContext]=\"{ $implicit: rowContextMenuPosition.value}\">\n        </ng-container>\n      </ul>\n      <ng-container *ngFor=\"let row of data | sort:sortBy | search:term | global:globalSearchTerm | paginate: { itemsPerPage: limit, currentPage: page, totalItems: count, id: id };\n              let rowIndex = index\">\n        <tr\n          (click)=\"onClick($event, row, '', null, rowIndex)\"\n          (contextmenu)=\"onRowContextMenu($event, row, '', null, rowIndex)\"\n          (dblclick)=\"onDoubleClick($event, row, '', null, rowIndex)\"\n          [class.ngx-table__table-row--selected]=\"rowIndex == selectedRow && !config.selectCell\">\n          <ng-container\n            [ngTemplateOutlet]=\"rowTemplate\"\n            [ngTemplateOutletContext]=\"{ $implicit: row, index: rowIndex }\">\n          </ng-container>\n          <td *ngIf=\"config.detailsTemplate\">\n            <span class=\"ngx-icon\"\n                  *ngIf=\"arrowDefinition\"\n                  [ngClass]=\"isRowCollapsed(rowIndex) ? 'ngx-icon-arrow-down' : 'ngx-icon-arrow-right'\"\n                  (click)=\"collapseRow(rowIndex)\">\n            </span>\n          </td>\n        </tr>\n        <tr\n          *ngIf=\"(config.detailsTemplate && selectedDetailsTemplateRowId.has(rowIndex)) || config.collapseAllRows\">\n          <td [attr.colspan]=\"columns.length + 1\">\n            <ng-container\n              [ngTemplateOutlet]=\"detailsTemplate\"\n              [ngTemplateOutletContext]=\"{ $implicit: row, index: rowIndex  }\">\n            </ng-container>\n          </td>\n        </tr>\n      </ng-container>\n    </ng-container>\n    <ng-container *ngIf=\"!rowTemplate && !config.groupRows\">\n      <ul class=\"ngx-table__table-row-context-menu\"\n          [ngStyle]=\"{'position': 'absolute', 'top': rowContextMenuPosition.top, 'left': rowContextMenuPosition.left }\"\n          *ngIf=\"rowContextMenuPosition.top\">\n        <ng-container\n          [ngTemplateOutlet]=\"rowContextMenu\"\n          [ngTemplateOutletContext]=\"{ $implicit: rowContextMenuPosition.value}\">\n        </ng-container>\n      </ul>\n      <ng-container\n        *ngFor=\"let row of data | sort:sortBy | search:term | global:globalSearchTerm | paginate: { itemsPerPage: limit, currentPage: page, totalItems: count, id: id };\n                  let rowIndex = index\">\n        <tr [class.ngx-table__table-row--selected]=\"rowIndex == selectedRow && !config.selectCell\">\n          <td *ngIf=\"config.checkboxes\">\n            <label class=\"ngx-form-checkbox\">\n              <input type=\"checkbox\"\n                     id=\"checkbox-{{rowIndex}}\"\n                     [checked]=\"isSelected\"\n                     (change)=\"onCheckboxSelect($event, row, rowIndex)\">\n              <i class=\"ngx-form-icon\"></i>\n            </label>\n          </td>\n          <ng-container *ngFor=\"let column of columns; let colIndex = index\">\n            <td (click)=\"onClick($event, row, column.key, colIndex, rowIndex)\"\n                (contextmenu)=\"onRowContextMenu($event, row, column.key, colIndex, rowIndex)\"\n                (dblclick)=\"onDoubleClick($event, row, column.key, colIndex, rowIndex)\"\n                [class.ngx-table__table-col--selected]=\"colIndex == selectedCol && !config.selectCell\"\n                [class.ngx-table__table-cell--selected]=\"colIndex == selectedCol && rowIndex == selectedRow && !config.selectCol && !config.selectRow\"\n            >\n              <div *ngIf=\"!column.cellTemplate\">{{ row | render:column.key }}</div>\n              <ng-container\n                *ngIf=\"column.cellTemplate\"\n                [ngTemplateOutlet]=\"column.cellTemplate\"\n                [ngTemplateOutletContext]=\"{ $implicit: row, rowIndex: rowIndex }\">\n              </ng-container>\n            </td>\n          </ng-container>\n          <td *ngIf=\"config.additionalActions || config.detailsTemplate\">\n            <span class=\"ngx-icon\"\n                  *ngIf=\"arrowDefinition\"\n                  [ngClass]=\"isRowCollapsed(rowIndex) ? 'ngx-icon-arrow-down' : 'ngx-icon-arrow-right'\"\n                  (click)=\"collapseRow(rowIndex)\">\n            </span>\n          </td>\n        </tr>\n        <tr\n          *ngIf=\"(config.detailsTemplate && selectedDetailsTemplateRowId.has(rowIndex)) || config.collapseAllRows\">\n          <td *ngIf=\"config.checkboxes\"></td>\n          <td [attr.colspan]=\"columns.length + 1\">\n            <ng-container\n              [ngTemplateOutlet]=\"detailsTemplate\"\n              [ngTemplateOutletContext]=\"{ $implicit: row, index: rowIndex }\">\n            </ng-container>\n          </td>\n        </tr>\n      </ng-container>\n    </ng-container>\n    <ng-container *ngIf=\"!rowTemplate && config.groupRows\">\n      <ng-container\n        *ngFor=\"let group of grouped | sort:sortBy:config | search:term:config | global:globalSearchTerm | paginate: { itemsPerPage: limit, currentPage: page, totalItems: count, id: id }; let rowIndex = index\">\n        <tr>\n          <ng-container *ngIf=\"!groupRowsHeaderTemplate\">\n            <td [attr.colspan]=\"columns.length\">\n              <div>{{group[0][groupRowsBy]}} ({{group.length}})</div>\n            </td>\n          </ng-container>\n          <ng-container\n            *ngIf=\"groupRowsHeaderTemplate\"\n            [ngTemplateOutlet]=\"groupRowsHeaderTemplate\"\n            [ngTemplateOutletContext]=\"{\n              total: group.length,\n              key: groupRowsBy,\n              value: group[0] ? group[0][groupRowsBy] : '',\n              group: group,\n              index: rowIndex\n            }\">\n          </ng-container>\n          <td>\n            <span class=\"ngx-icon\"\n                  *ngIf=\"arrowDefinition\"\n                  [ngClass]=\"isRowCollapsed(rowIndex) ? 'ngx-icon-arrow-down' : 'ngx-icon-arrow-right'\"\n                  (click)=\"collapseRow(rowIndex)\">\n            </span>\n          </td>\n        </tr>\n        <ng-container *ngIf=\"selectedDetailsTemplateRowId.has(rowIndex)\">\n          <tr *ngFor=\"let row of group\">\n            <td *ngFor=\"let column of columns\">\n              {{ row | render:column.key }}\n              <!-- TODO allow users to add groupRowsTemplateRef -->\n            </td>\n            <td></td>\n          </tr>\n        </ng-container>\n      </ng-container>\n    </ng-container>\n    </tbody>\n    <tbody *ngIf=\"data && !config.isLoading && config.draggable\" cdkDropList (cdkDropListDropped)=\"onDrop($event)\">\n    <ng-container *ngIf=\"!rowTemplate && !config.groupRows\">\n      <ng-container\n        *ngFor=\"let row of data | sort:sortBy | search:term | global:globalSearchTerm | paginate: { itemsPerPage: limit, currentPage: page, totalItems: count, id: id };\n                  let rowIndex = index\">\n        <tr class=\"ngx-draggable-row\" cdkDrag>\n          <td *ngIf=\"config.checkboxes\">\n            <label class=\"ngx-form-checkbox\">\n              <input type=\"checkbox\"\n                     id=\"checkbox-draggable-{{rowIndex}}\"\n                     [checked]=\"isSelected\"\n                     (change)=\"onCheckboxSelect($event, row, rowIndex)\">\n              <i class=\"ngx-form-icon\"></i>\n            </label>\n          </td>\n          <ng-container *ngFor=\"let column of columns; let colIndex = index\">\n            <td (click)=\"onClick($event, row, column.key, colIndex, rowIndex)\"\n                (dblclick)=\"onDoubleClick($event, row, column.key, colIndex, rowIndex)\"\n                [class.ngx-table__table-col--selected]=\"colIndex == selectedCol && !config.selectCell\"\n                [class.ngx-table__table-cell--selected]=\"colIndex == selectedCol && rowIndex == selectedRow && !config.selectCol && !config.selectRow\"\n            >\n              <div>{{ row | render:column.key }}</div>\n            </td>\n          </ng-container>\n        </tr>\n      </ng-container>\n    </ng-container>\n    </tbody>\n    <tbody *ngIf=\"!isLoading && (!data || (data && data.length < 1))\">\n    <tr class=\"ngx-table__body-empty\">\n      <ng-container\n        *ngIf=\"noResultsTemplate\"\n        [ngTemplateOutlet]=\"noResultsTemplate\">\n      </ng-container>\n      <td [attr.colspan]=\"columns && columns.length + 1\" *ngIf=\"!noResultsTemplate\">\n        <div class=\"ngx-table__table-no-results\">\n          No results\n        </div>\n      </td>\n    </tr>\n    </tbody>\n    <tbody *ngIf=\"isLoading\">\n    <tr class=\"ngx-table__body-loading\">\n      <td [attr.colspan]=\"columns && columns.length + 1\">\n        <div [style.height]=\"loadingHeight\"\n             class=\"ngx-table__table-loader-wrapper\">\n          <div class=\"ngx-table__table-loader\"></div>\n        </div>\n      </td>\n    </tr>\n    </tbody>\n    <tfoot *ngIf=\"summaryTemplate\">\n    <tr>\n      <ng-container\n        [ngTemplateOutlet]=\"summaryTemplate\"\n        [ngTemplateOutletContext]=\"{ total: data.length, limit: limit, page: page  }\">\n      </ng-container>\n    </tr>\n    </tfoot>\n  </table>\n  <pagination\n    [attr.id]=\"'pagination' + id\"\n    *ngIf=\"config.paginationEnabled\"\n    [id]=\"id\"\n    #paginationComponent\n    [config]=\"config\"\n    [pagination]=\"pagination\"\n    (updateRange)=\"onPagination($event)\">\n  </pagination>\n</div>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        BaseComponent.ctorParameters = function () {
            return [
                { type: core.ChangeDetectorRef }
            ];
        };
        BaseComponent.propDecorators = {
            paginationComponent: [{ type: core.ViewChild, args: ['paginationComponent',] }],
            th: [{ type: core.ViewChild, args: ['th',] }],
            configuration: [{ type: core.Input, args: ['configuration',] }],
            data: [{ type: core.Input }],
            pagination: [{ type: core.Input }],
            groupRowsBy: [{ type: core.Input }],
            id: [{ type: core.Input }],
            toggleRowIndex: [{ type: core.Input }],
            detailsTemplate: [{ type: core.Input }],
            summaryTemplate: [{ type: core.Input }],
            groupRowsHeaderTemplate: [{ type: core.Input }],
            filtersTemplate: [{ type: core.Input }],
            selectAllTemplate: [{ type: core.Input }],
            noResultsTemplate: [{ type: core.Input }],
            rowContextMenu: [{ type: core.Input }],
            columns: [{ type: core.Input }],
            event: [{ type: core.Output }],
            rowTemplate: [{ type: core.ContentChild, args: [core.TemplateRef,] }]
        };
        return BaseComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var CsvExportComponent = /** @class */ (function () {
        function CsvExportComponent() {
        }
        /**
         * @return {?}
         */
        CsvExportComponent.prototype.exportCsv = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var data = this.data;
                /** @type {?} */
                var csvContent = 'data:text/csv;charset=utf-8,';
                /** @type {?} */
                var dataString = '';
                /** @type {?} */
                var x = [];
                /** @type {?} */
                var keys = Object.keys(this.data[0]);
                data.forEach(( /**
                 * @param {?} row
                 * @param {?} index
                 * @return {?}
                 */function (row, index) {
                    x[index] = [];
                    keys.forEach(( /**
                     * @param {?} i
                     * @return {?}
                     */function (i) {
                        if (row.hasOwnProperty(i)) {
                            if (typeof row[i] === 'object') {
                                row[i] = 'Object'; // so far just change object to "Object" string
                            }
                            x[index].push(row[i]);
                        }
                    }));
                }));
                csvContent += keys + '\n';
                x.forEach(( /**
                 * @param {?} row
                 * @param {?} index
                 * @return {?}
                 */function (row, index) {
                    dataString = row.join(',');
                    csvContent += index < data.length ? dataString + '\n' : dataString;
                }));
                /** @type {?} */
                var encodedUri = encodeURI(csvContent);
                /** @type {?} */
                var link = document.createElement('a');
                link.setAttribute('href', encodedUri);
                link.setAttribute('download', 'my_data.csv');
                link.click();
            };
        CsvExportComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'app-csv-export',
                        template: "\n    <a (click)=\"exportCsv()\" class=\"ngx-menu-item\">\n      CSV export\n    </a>"
                    }] }
        ];
        CsvExportComponent.propDecorators = {
            data: [{ type: core.Input }]
        };
        return CsvExportComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var HeaderComponent = /** @class */ (function () {
        function HeaderComponent() {
            this.update = new core.EventEmitter();
        }
        /**
         * @param {?} key
         * @return {?}
         */
        HeaderComponent.prototype.unifyKey = /**
         * @param {?} key
         * @return {?}
         */
            function (key) {
                return key.replace('.', '_');
            };
        /**
         * @param {?} input
         * @return {?}
         */
        HeaderComponent.prototype.onSearch = /**
         * @param {?} input
         * @return {?}
         */
            function (input) {
                this.update.emit([
                    { value: input.value, key: this.column.key },
                ]);
            };
        HeaderComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'table-header',
                        template: "<label for=\"search_{{ unifyKey(column.key) }}\">\n  <input type=\"text\"\n         id=\"search_{{ unifyKey(column.key) }}\"\n         aria-label=\"Search\"\n         placeholder=\"{{ column.placeholder ? column.placeholder : column.title }}\"\n         class=\"ngx-table__header-search\"\n         #input\n         (input)=\"onSearch(input)\"\n  >\n</label>\n"
                    }] }
        ];
        HeaderComponent.propDecorators = {
            column: [{ type: core.Input }],
            update: [{ type: core.Output }]
        };
        return HeaderComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var PaginationComponent = /** @class */ (function () {
        function PaginationComponent() {
            this.updateRange = new core.EventEmitter();
            this.ranges = [5, 10, 25, 50, 100];
            this.limit = DefaultConfigService.config.rows;
            this.showRange = false;
            this.screenReaderPaginationLabel = 'Pagination';
            this.screenReaderPageLabel = 'page';
            this.screenReaderCurrentLabel = 'You are on page';
            this.previousLabel = '';
            this.nextLabel = '';
            this.directionLinks = true;
        }
        /**
         * @param {?} page
         * @return {?}
         */
        PaginationComponent.prototype.onPageChange = /**
         * @param {?} page
         * @return {?}
         */
            function (page) {
                this.updateRange.emit({
                    page: page,
                    limit: this.limit,
                });
            };
        /**
         * @param {?} limit
         * @param {?} callFromAPI
         * @return {?}
         */
        PaginationComponent.prototype.changeLimit = /**
         * @param {?} limit
         * @param {?} callFromAPI
         * @return {?}
         */
            function (limit, callFromAPI) {
                if (!callFromAPI) {
                    this.showRange = !this.showRange;
                }
                this.limit = limit;
                this.updateRange.emit({
                    page: 1,
                    limit: limit,
                });
            };
        PaginationComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'pagination',
                        template: "<div class=\"ngx-pagination-wrapper\"\n     [class.ngx-table__table--dark-pagination-wrapper]=\"config.tableLayout.theme === 'dark'\">\n  <div class=\"ngx-pagination-steps\">\n    <pagination-template\n      #paginationDirective=\"paginationApi\"\n      id=\"pagination-controls\"\n      [id]=\"id\"\n      [class.ngx-table__table--dark-pagination]=\"config.tableLayout.theme === 'dark'\"\n      [maxSize]=\"config.paginationMaxSize || 5\"\n      (pageChange)=\"onPageChange($event)\">\n      <ul class=\"ngx-pagination\"\n          role=\"navigation\"\n          [attr.aria-label]=\"screenReaderPaginationLabel\"\n          [class.responsive]=\"true\">\n        <li class=\"pagination-previous\" [class.disabled]=\"paginationDirective.isFirstPage()\" *ngIf=\"directionLinks\">\n          <a tabindex=\"0\" *ngIf=\"1 < paginationDirective.getCurrent()\" (keyup.enter)=\"paginationDirective.previous()\"\n             (click)=\"paginationDirective.previous()\"\n             [attr.aria-label]=\"previousLabel + ' ' + screenReaderPageLabel\">\n            {{ previousLabel }} <span class=\"show-for-sr\">{{ screenReaderPageLabel }}</span>\n          </a>\n          <span *ngIf=\"paginationDirective.isFirstPage()\">\n                {{ previousLabel }} <span class=\"show-for-sr\">{{ screenReaderPageLabel }}</span>\n            </span>\n        </li>\n        <li class=\"small-screen\">\n          {{ paginationDirective.getCurrent() }} / {{ paginationDirective.getLastPage() }}\n        </li>\n        <li [class.current]=\"paginationDirective.getCurrent() === page.value\"\n            [class.ellipsis]=\"page.label === '...'\"\n            *ngFor=\"let page of paginationDirective.pages\">\n          <a tabindex=\"0\" (keyup.enter)=\"paginationDirective.setCurrent(page.value)\"\n             (click)=\"paginationDirective.setCurrent(page.value)\"\n             *ngIf=\"paginationDirective.getCurrent() !== page.value\">\n            <span class=\"show-for-sr\">{{ screenReaderPageLabel }} </span>\n            <span>{{ page.label }}</span>\n          </a>\n          <ng-container *ngIf=\"paginationDirective.getCurrent() === page.value\">\n            <span class=\"show-for-sr\">{{ screenReaderCurrentLabel }} </span>\n            <span>{{ page.label }}</span>\n          </ng-container>\n        </li>\n        <li class=\"pagination-next\" [class.disabled]=\"paginationDirective.isLastPage()\" *ngIf=\"directionLinks\">\n          <a tabindex=\"0\" *ngIf=\"!paginationDirective.isLastPage()\" (keyup.enter)=\"paginationDirective.next()\"\n             (click)=\"paginationDirective.next()\"\n             [attr.aria-label]=\"nextLabel + ' ' + screenReaderPageLabel\">\n            {{ nextLabel }} <span class=\"show-for-sr\">{{ screenReaderPageLabel }}</span>\n          </a>\n          <span *ngIf=\"paginationDirective.isLastPage()\">\n                 {{ nextLabel }} <span class=\"show-for-sr\">{{ screenReaderPageLabel }}</span>\n            </span>\n        </li>\n      </ul>\n    </pagination-template>\n  </div>\n  <div class=\"ngx-pagination-range\"\n       [class.ngx-table__table--dark-pagination-range]=\"config.tableLayout.theme === 'dark'\"\n       *ngIf=\"config.paginationRangeEnabled\">\n    <div class=\"ngx-dropdown ngx-pagination-range-dropdown\"\n         [class.ngx-active]=\"showRange\"\n         id=\"rowAmount\">\n      <div class=\"ngx-btn-group\">\n        <div class=\"ngx-pagination-range-dropdown-button\"\n             (click)=\"showRange = !showRange\">\n          {{limit}} <i class=\"ngx-icon ngx-icon-arrow-down\"></i>\n        </div>\n        <ul class=\"ngx-menu\" *ngIf=\"showRange\">\n          <li class=\"ngx-pagination-range-dropdown-button-item\"\n              (click)=\"changeLimit(limit, false)\"\n              *ngFor=\"let limit of ranges\">\n            <span>{{limit}}</span>\n          </li>\n        </ul>\n      </div>\n    </div>\n  </div>\n</div>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        PaginationComponent.propDecorators = {
            paginationDirective: [{ type: core.ViewChild, args: ['paginationDirective',] }],
            pagination: [{ type: core.Input }],
            config: [{ type: core.Input }],
            id: [{ type: core.Input }],
            updateRange: [{ type: core.Output }]
        };
        return PaginationComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var GlobalSearchPipe = /** @class */ (function () {
        function GlobalSearchPipe() {
        }
        /**
         * @param {?} dataArr
         * @param {?} filter
         * @return {?}
         */
        GlobalSearchPipe.prototype.transform = /**
         * @param {?} dataArr
         * @param {?} filter
         * @return {?}
         */
            function (dataArr, filter) {
                if (typeof dataArr === 'undefined') {
                    return;
                }
                if (typeof filter === 'undefined' || Object.keys(filter).length === 0 || filter === '') {
                    return dataArr;
                }
                return dataArr.filter(( /**
                 * @param {?} row
                 * @return {?}
                 */function (row) {
                    /** @type {?} */
                    var element = JSON.stringify(Object.values(row));
                    return element.toLocaleLowerCase().indexOf(filter.toLocaleLowerCase()) !== -1;
                }));
            };
        GlobalSearchPipe.decorators = [
            { type: core.Pipe, args: [{
                        name: 'global',
                    },] }
        ];
        return GlobalSearchPipe;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var FiltersService = /** @class */ (function () {
        function FiltersService() {
        }
        /**
         * @param {?} p
         * @param {?} o
         * @return {?}
         */
        FiltersService.getPath = /**
         * @param {?} p
         * @param {?} o
         * @return {?}
         */
            function (p, o) {
                // https://github.com/dherges/ng-packagr/issues/696
                /* tslint:disable-next-line */
                /** @type {?} */
                var result = p.reduce(( /**
                 * @param {?} xs
                 * @param {?} x
                 * @return {?}
                 */function (xs, x) { return (xs && typeof xs[x] !== 'undefined') ? xs[x] : null; }), o);
                return result;
            };
        FiltersService.decorators = [
            { type: core.Injectable }
        ];
        return FiltersService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var RenderPipe = /** @class */ (function () {
        function RenderPipe() {
        }
        /**
         * @param {?} row
         * @param {?} key
         * @return {?}
         */
        RenderPipe.prototype.transform = /**
         * @param {?} row
         * @param {?} key
         * @return {?}
         */
            function (row, key) {
                /** @type {?} */
                var split = key.split('.');
                return FiltersService.getPath(split, row);
            };
        RenderPipe.decorators = [
            { type: core.Pipe, args: [{
                        name: 'render',
                    },] }
        ];
        return RenderPipe;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SearchPipe = /** @class */ (function () {
        function SearchPipe() {
            this.filters = {};
        }
        /**
         * @param {?} array
         * @param {?=} filter
         * @param {?=} config
         * @return {?}
         */
        SearchPipe.prototype.transform = /**
         * @param {?} array
         * @param {?=} filter
         * @param {?=} config
         * @return {?}
         */
            function (array, filter, config) {
                var _this = this;
                if (typeof array === 'undefined') {
                    return;
                }
                if (typeof filter === 'undefined') {
                    return array;
                }
                filter.forEach(( /**
                 * @param {?} f
                 * @return {?}
                 */function (f) {
                    _this.filters[f.key] = f.value.toString().toLocaleLowerCase();
                    if (Object.keys(f).length === 0 || f.value === '') {
                        delete _this.filters[f.key];
                    }
                }));
                if (config && config.groupRows) {
                    return array.map(( /**
                     * @param {?} arr
                     * @return {?}
                     */function (arr) { return _this.filterGroup(arr); }));
                }
                return this.filterGroup(array);
            };
        /**
         * @private
         * @param {?} array
         * @return {?}
         */
        SearchPipe.prototype.filterGroup = /**
         * @private
         * @param {?} array
         * @return {?}
         */
            function (array) {
                var _this = this;
                return array.filter(( /**
                 * @param {?} obj
                 * @return {?}
                 */function (obj) {
                    return Object.keys(_this.filters).every(( /**
                     * @param {?} c
                     * @return {?}
                     */function (c) {
                        /** @type {?} */
                        var split = c.split('.');
                        /** @type {?} */
                        var val = FiltersService.getPath(split, obj);
                        /** @type {?} */
                        var element = (typeof val === 'object') ? JSON.stringify(val) : val.toString().toLocaleLowerCase();
                        /** @type {?} */
                        var strings = _this.filters[c].split(',');
                        return strings.some(( /**
                         * @param {?} string
                         * @return {?}
                         */function (string) { return element.indexOf(string.trim()) > -1; }));
                    }));
                }));
            };
        SearchPipe.decorators = [
            { type: core.Pipe, args: [{
                        name: 'search',
                    },] }
        ];
        return SearchPipe;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var SortPipe = /** @class */ (function () {
        function SortPipe() {
            this.defaultArray = [];
        }
        /**
         * @private
         * @param {?} aV
         * @param {?} bV
         * @return {?}
         */
        SortPipe.isNaN = /**
         * @private
         * @param {?} aV
         * @param {?} bV
         * @return {?}
         */
            function (aV, bV) {
                return (isNaN(parseFloat(aV)) || !isFinite(aV)) || (isNaN(parseFloat(bV)) || !isFinite(bV));
            };
        /**
         * @private
         * @param {?} a
         * @param {?} b
         * @param {?} key
         * @return {?}
         */
        SortPipe.compare = /**
         * @private
         * @param {?} a
         * @param {?} b
         * @param {?} key
         * @return {?}
         */
            function (a, b, key) {
                /** @type {?} */
                var split = key.split('.');
                /** @type {?} */
                var aPath = FiltersService.getPath(split, a);
                /** @type {?} */
                var bPath = FiltersService.getPath(split, b);
                /** @type {?} */
                var aValue = (aPath + '').toLowerCase();
                /** @type {?} */
                var bValue = (bPath + '').toLowerCase();
                if (SortPipe.isNaN(aPath, bPath)) {
                    return aValue.localeCompare(bValue);
                }
                else {
                    if (parseFloat(aPath) < parseFloat(bPath)) {
                        return -1;
                    }
                    if (parseFloat(aPath) > parseFloat(bPath)) {
                        return 1;
                    }
                }
                return 0;
            };
        /**
         * @param {?} array
         * @param {?=} filter
         * @param {?=} config
         * @return {?}
         */
        SortPipe.prototype.transform = /**
         * @param {?} array
         * @param {?=} filter
         * @param {?=} config
         * @return {?}
         */
            function (array, filter, config) {
                if (this.defaultArray.length === 0) {
                    this.defaultArray = array;
                }
                if (!filter.key || filter.key === '') {
                    return array;
                }
                if (filter.order === '') {
                    return this.defaultArray;
                }
                if (filter.order === 'asc') {
                    return this.sortAsc(config, array, filter);
                }
                else {
                    return this.sortDesc(config, array, filter);
                }
            };
        /**
         * @private
         * @param {?} config
         * @param {?} array
         * @param {?} filter
         * @return {?}
         */
        SortPipe.prototype.sortAsc = /**
         * @private
         * @param {?} config
         * @param {?} array
         * @param {?} filter
         * @return {?}
         */
            function (config, array, filter) {
                if (config && config.groupRows) {
                    return array.map(( /**
                     * @param {?} arr
                     * @return {?}
                     */function (arr) {
                        return arr.sort(( /**
                         * @param {?} a
                         * @param {?} b
                         * @return {?}
                         */function (a, b) { return SortPipe.compare(a, b, filter.key); }));
                    }));
                }
                return array.sort(( /**
                 * @param {?} a
                 * @param {?} b
                 * @return {?}
                 */function (a, b) { return SortPipe.compare(a, b, filter.key); }));
            };
        /**
         * @private
         * @param {?} config
         * @param {?} array
         * @param {?} filter
         * @return {?}
         */
        SortPipe.prototype.sortDesc = /**
         * @private
         * @param {?} config
         * @param {?} array
         * @param {?} filter
         * @return {?}
         */
            function (config, array, filter) {
                if (config && config.groupRows) {
                    return array.map(( /**
                     * @param {?} arr
                     * @return {?}
                     */function (arr) {
                        return arr.sort(( /**
                         * @param {?} a
                         * @param {?} b
                         * @return {?}
                         */function (a, b) { return SortPipe.compare(b, a, filter.key); }));
                    }));
                }
                return array.sort(( /**
                 * @param {?} a
                 * @param {?} b
                 * @return {?}
                 */function (a, b) { return SortPipe.compare(b, a, filter.key); }));
            };
        SortPipe.decorators = [
            { type: core.Pipe, args: [{
                        name: 'sort',
                    },] }
        ];
        return SortPipe;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BaseModule = /** @class */ (function () {
        function BaseModule() {
        }
        BaseModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [
                            BaseComponent,
                            CsvExportComponent,
                            HeaderComponent,
                            PaginationComponent,
                            SearchPipe,
                            RenderPipe,
                            GlobalSearchPipe,
                            SortPipe,
                        ],
                        imports: [
                            common.CommonModule,
                            ngxPagination.NgxPaginationModule,
                            dragDrop.DragDropModule,
                        ],
                        exports: [BaseComponent],
                    },] }
        ];
        return BaseModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var TableModule = /** @class */ (function () {
        function TableModule() {
        }
        TableModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            BaseModule,
                        ],
                        bootstrap: [BaseComponent],
                        exports: [BaseComponent],
                        providers: [],
                    },] }
        ];
        return TableModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.DefaultConfig = DefaultConfig;
    exports.TableModule = TableModule;
    exports.STYLE = STYLE;
    exports.THEME = THEME;
    exports.Event = Event;
    exports.API = API;
    exports.c = BaseComponent;
    exports.b = BaseModule;
    exports.f = CsvExportComponent;
    exports.g = HeaderComponent;
    exports.h = PaginationComponent;
    exports.k = GlobalSearchPipe;
    exports.j = RenderPipe;
    exports.i = SearchPipe;
    exports.l = SortPipe;
    exports.a = DefaultConfigService;
    exports.d = GroupRowsService;
    exports.e = StyleService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
