// ngx-easy-table/esm5/lib/model/api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var API = {
    rowContextMenuClicked: 'rowContextMenuClicked',
    setInputValue: 'setInputValue',
    toolPanelClicked: 'toolPanelClicked',
    onGlobalSearch: 'onGlobalSearch',
    setPaginationCurrentPage: 'setPaginationCurrentPage',
    getPaginationCurrentPage: 'getPaginationCurrentPage',
    getPaginationTotalItems: 'getPaginationTotalItems',
    getPaginationLastPage: 'getPaginationLastPage',
    setPaginationRange: 'setPaginationRange',
    setPaginationPreviousLabel: 'setPaginationPreviousLabel',
    setPaginationNextLabel: 'setPaginationNextLabel',
    setPaginationDisplayLimit: 'setPaginationDisplayLimit',
    setTableClass: 'setTableClass',
    setRowClass: 'setRowClass',
    setCellClass: 'setCellClass',
    setRowStyle: 'setRowStyle',
    setCellStyle: 'setCellStyle',
    sortBy: 'sortBy',
};
export { API };
/**
 * @record
 */
export function APIDefinition() { }
if (false) {
    /**
     * @param {?} event
     * @return {?}
     */
    APIDefinition.prototype.apiEvent = function (event) { };
}
