// ngx-easy-table/esm5/lib/model/config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Config() { }
if (false) {
    /** @type {?} */
    Config.prototype.searchEnabled;
    /** @type {?} */
    Config.prototype.headerEnabled;
    /** @type {?} */
    Config.prototype.orderEnabled;
    /** @type {?|undefined} */
    Config.prototype.orderEventOnly;
    /** @type {?} */
    Config.prototype.paginationEnabled;
    /** @type {?} */
    Config.prototype.exportEnabled;
    /** @type {?} */
    Config.prototype.clickEvent;
    /** @type {?} */
    Config.prototype.selectRow;
    /** @type {?} */
    Config.prototype.selectCol;
    /** @type {?} */
    Config.prototype.selectCell;
    /** @type {?} */
    Config.prototype.rows;
    /** @type {?} */
    Config.prototype.additionalActions;
    /** @type {?} */
    Config.prototype.serverPagination;
    /** @type {?} */
    Config.prototype.isLoading;
    /** @type {?} */
    Config.prototype.detailsTemplate;
    /** @type {?} */
    Config.prototype.groupRows;
    /** @type {?} */
    Config.prototype.paginationRangeEnabled;
    /** @type {?} */
    Config.prototype.collapseAllRows;
    /** @type {?} */
    Config.prototype.checkboxes;
    /** @type {?} */
    Config.prototype.resizeColumn;
    /** @type {?} */
    Config.prototype.fixedColumnWidth;
    /** @type {?} */
    Config.prototype.horizontalScroll;
    /** @type {?} */
    Config.prototype.draggable;
    /** @type {?} */
    Config.prototype.logger;
    /** @type {?|undefined} */
    Config.prototype.showDetailsArrow;
    /** @type {?|undefined} */
    Config.prototype.showContextMenu;
    /** @type {?|undefined} */
    Config.prototype.persistState;
    /** @type {?|undefined} */
    Config.prototype.paginationMaxSize;
    /** @type {?|undefined} */
    Config.prototype.threeWaySort;
    /** @type {?} */
    Config.prototype.tableLayout;
}
/** @enum {string} */
var STYLE = {
    TINY: 'tiny',
    BIG: 'big',
    NORMAL: 'normal',
};
export { STYLE };
/** @enum {string} */
var THEME = {
    LIGHT: 'light',
    DARK: 'dark',
};
export { THEME };
