// @syncfusion/ej2-richtexteditor/src/rich-text-editor/formatter/formatter.js
import { extend, isNullOrUndefined, Browser } from '@syncfusion/ej2-base';
import * as CONSTANT from '../base/constant';
import { updateUndoRedoStatus, isIDevice } from '../base/util';
import { KEY_DOWN, KEY_UP } from './../../common/constant';
/**
 * Formatter
 * @hidden
 */
var Formatter = /** @class */ (function () {
    function Formatter() {
    }
    /**
     * To execute the command
     * @param  {IRichTextEditor} self
     * @param  {ActionBeginEventArgs} args
     * @param  {MouseEvent|KeyboardEvent} event
     * @param  {NotifyArgs} value
     */
    Formatter.prototype.process = function (self, args, event, value) {
        var selection = self.contentModule.getDocument().getSelection();
        var range = (selection.rangeCount > 0) ? selection.getRangeAt(selection.rangeCount - 1) : null;
        if (!isNullOrUndefined(args)
            && args.item.command
            && args.item.command !== 'Table'
            && args.item.command !== 'Actions'
            && args.item.command !== 'Links'
            && args.item.command !== 'Images'
            && range
            && !(self.contentModule.getEditPanel().contains(this.getAncestorNode(range.commonAncestorContainer))
                || self.contentModule.getEditPanel() === range.commonAncestorContainer
                || self.contentModule.getPanel() === range.commonAncestorContainer)) {
            return;
        }
        if (isNullOrUndefined(args)) {
            var action = event.action;
            if (action !== 'tab' && action !== 'enter' && action !== 'space' && action !== 'escape') {
                args = {};
                if (self.editorMode === 'Markdown' && action === 'insert-table') {
                    value = {
                        'headingText': self.localeObj.getConstant('TableHeadingText'),
                        'colText': self.localeObj.getConstant('TableColText')
                    };
                }
                var items = {
                    originalEvent: event, cancel: false,
                    requestType: action || (event.key + 'Key'),
                    itemCollection: value
                };
                extend(args, args, items, true);
                self.trigger(CONSTANT.actionBegin, args);
                if (args.cancel) {
                    if (action === 'paste' || action === 'cut' || action === 'copy') {
                        event.preventDefault();
                    }
                    return;
                }
            }
            this.editorManager.observer.notify((event.type === 'keydown' ? KEY_DOWN : KEY_UP), {
                event: event,
                callBack: this.onSuccess.bind(this, self),
                value: value
            });
        }
        else if (!isNullOrUndefined(args) && args.item.command && args.item.subCommand && ((args.item.command !== args.item.subCommand
            && args.item.command !== 'Font')
            || ((args.item.subCommand === 'FontName' || args.item.subCommand === 'FontSize') && args.name === 'dropDownSelect')
            || ((args.item.subCommand === 'BackgroundColor' || args.item.subCommand === 'FontColor')
                && args.name === 'colorPickerChanged'))) {
            extend(args, args, { requestType: args.item.subCommand, cancel: false, itemCollection: value }, true);
            self.trigger(CONSTANT.actionBegin, args);
            if (args.cancel) {
                return;
            }
            if (this.getUndoRedoStack().length === 0 && args.item.command !== 'Links' && args.item.command !== 'Images') {
                this.saveData();
            }
            self.isBlur = false;
            self.contentModule.getEditPanel().focus();
            var command = args.item.subCommand.toLocaleLowerCase();
            if (command === 'paste' || command === 'cut' || command === 'copy') {
                self.clipboardAction(command, event);
            }
            else {
                this.editorManager.observer.notify(CONSTANT.checkUndo, { subCommand: args.item.subCommand });
                this.editorManager.execCommand(args.item.command, args.item.subCommand, event, this.onSuccess.bind(this, self), args.item.value, value, ('#' + self.getID() + ' iframe'));
            }
        }
        if (isNullOrUndefined(event) || event && event.action !== 'copy') {
            this.enableUndo(self);
        }
    };
    Formatter.prototype.getAncestorNode = function (node) {
        node = node.nodeType === 3 ? node.parentNode : node;
        return node;
    };
    Formatter.prototype.onKeyHandler = function (self, e) {
        var _this = this;
        this.editorManager.observer.notify(KEY_UP, {
            event: e, callBack: function () {
                _this.enableUndo(self);
            }
        });
    };
    Formatter.prototype.onSuccess = function (self, events) {
        if (isNullOrUndefined(events.event) || (events && events.event.action !== 'copy')) {
            this.enableUndo(self);
            self.notify(CONSTANT.execCommandCallBack, events);
        }
        self.trigger(CONSTANT.actionComplete, events);
        if (events.requestType === 'Images' || events.requestType === 'Links' && self.editorMode === 'HTML') {
            var args = events;
            if (events.requestType === 'Links' && events.event &&
                events.event.type === 'keydown' &&
                events.event.keyCode === 32) {
                return;
            }
            self.notify(CONSTANT.insertCompleted, {
                args: args.event, type: events.requestType, isNotify: true,
                elements: args.elements
            });
        }
        self.autoResize();
    };
    /**
     * Save the data for undo and redo action.
     */
    Formatter.prototype.saveData = function (e) {
        this.editorManager.undoRedoManager.saveData(e);
    };
    Formatter.prototype.getUndoStatus = function () {
        return this.editorManager.undoRedoManager.getUndoStatus();
    };
    Formatter.prototype.getUndoRedoStack = function () {
        return this.editorManager.undoRedoManager.undoRedoStack;
    };
    Formatter.prototype.enableUndo = function (self) {
        var status = this.getUndoStatus();
        if (self.inlineMode.enable && (!Browser.isDevice || isIDevice())) {
            updateUndoRedoStatus(self.quickToolbarModule.inlineQTBar.quickTBarObj, status);
        }
        else {
            if (self.toolbarModule) {
                updateUndoRedoStatus(self.toolbarModule.baseToolbar, status);
            }
        }
    };
    Formatter.prototype.undoRedoRefresh = function (iRichTextEditor) {
        if (this.editorManager.undoRedoManager.undoRedoStack.length) {
            this.editorManager.undoRedoManager.undoRedoStack = [];
            this.editorManager.undoRedoManager.steps = 0;
            iRichTextEditor.disableToolbarItem(['Undo', 'Redo']);
        }
    };
    return Formatter;
}());
export { Formatter };
